src/wayback/_http.py

```diff
@@ -0,0 +1,591 @@
+        The rate limits that should be applied to different URL paths. The keys
+        are URL path prefixes, e.g. ``"/cdx/search/"``, and the values are
+        :class:`wayback.RateLimit` objects. When requests are made, the rate
+        limit from the most specific matching path is used and execution will
+        pause to ensure the rate limit is not exceeded.
+
+
+    Examples
+    --------
+
+    Usage via multiple inheritance:
+
+    >>> class CombinedAdapter(RetryAndRateLimitAdapter, SomeActualHttpAdapterWithoutRateLimits):
+    >>>     ...
+
+    Usage via override:
+
+    >>> class MyHttpAdapter(RetryAndRateLimitAdapter):
+    >>>     def request_raw(
+    >>>         self,
+    >>>         method: str,
+    >>>         url: str,
+    >>>         *,
+    >>>         params: dict = None,
+    >>>         headers: dict = None,
+    >>>         follow_redirects: bool = True,
+    >>>         timeout: Union[int, Tuple[int, int]] = None
+    >>>     ) -> WaybackHttpResponse:
+    >>>         response = urllib.urlopen(...)
+    >>>         return make_response_from_urllib(response)
+    """
+
+    rate_limits: Dict[str, RateLimit]
+
+    # It seems Wayback sometimes produces 500 errors for transient issues, so
+    # they make sense to retry here. Usually not in other contexts, though.
+    retryable_statuses = frozenset((413, 421, 500, 502, 503, 504, 599))
+
+    # XXX: Some of these are requests-specific and should move WaybackRequestsAdapter.
+    retryable_errors = (ConnectTimeoutError, MaxRetryError, ReadTimeoutError,
+                        ProxyError, RetryError, Timeout)
+    # Handleable errors *may* be retryable, but need additional logic beyond
+    # just the error type. See `should_retry_error()`.
+    handleable_errors = (ConnectionError,) + retryable_errors
+
+    def __init__(
+        self,
+        retries: int = 6,
+        backoff: Real = 2,
+        rate_limits: Dict[str, RateLimit] = {}
+    ) -> None:
+        super().__init__()
+        self.retries = retries
+        self.backoff = backoff
+        # Sort rate limits by longest path first, so we always match the most
+        # specific path when looking for the right rate limit on any given URL.
+        self.rate_limits = {path: rate_limits[path]
+                            for path in sorted(rate_limits.keys(),
+                                               key=lambda k: len(k),
+                                               reverse=True)}
+
+    # The implementation of different features is split up by method here, so
+    # `request()` calls down through a stack of overrides:
+    #
+    #   request                        (ensure valid types/values/etc.)
+    #   └> _request_redirectable       (handle redirects)
+    #      └> _request_retryable       (handle retries for errors)
+    #         └> _request_rate_limited (handle rate limiting/throttling)
+    #            └> request_raw        (handle actual HTTP)
+    def request(
```

Comment:
This breaks the class up and makes its definition longer, but hopefully it makes it a little more manageable than the way things were done in `WaybackSession`.

These could also be implemented as totally separate adapters, which might be cleaner? But it felt simpler to just keep them together for now.

This was also partly influenced by my realization that we have a bug where we don’t implement our own redirect following, which means redirects don’t get rate limited (sigh). It this gives us a good place to slide *that* complexity in later (`_request_redirectable()` is a pass-through in this PR).